File: mergeRecursion/merge.js
function mergeSort(arr) {
  // base case
  if (arr.length < 1) {
    return arr;
  }

  // split array into two halves
  const middle = Math.floor(arr.length / 2);
  const leftHalf = arr.slice(0, middle);
  const rightHalf = arr.slice(middle);

  // recursively sort each half
  const sortedLeft = mergeSort(leftHalf);
  const sortedRight = mergeSort(rightHalf);

  // merge the sorted halves
  const mergedArr = merge(sortedLeft, sortedRight);
  return mergedArr;
}

function merge(left, right) {
  const mergedArr = [];
  let leftIdx = 0;
  let rightIdx = 0;

  // merge two halves into one sorted array
  while (leftIdx < left.length && rightIdx < right.length) {
    if (left[leftIdx] < right[rightIdx]) {
      mergedArr.push(left[leftIdx]);
      leftIdx++;
    } else {
      mergedArr.push(right[rightIdx]);
      rightIdx++;
    }
  }

  // add remaining elements to merged array
  mergedArr.push(...left.slice(leftIdx));
  mergedArr.push(...right.slice(rightIdx));

  return mergedArr;
}

File: mergeRecursion/merge.py
def merge_sort(arr):
    # Base case
    if len(arr) <= 1:
        return arr
    # divide array in two halves
    mid = len(arr) // 2
    left_half = merge_sort(arr[:mid])
    right_half = merge_sort(arr[mid:])
    # recursively sort each half
    left_half = merge_sort(left_half)
    right_half = merge_sort(right_half)

    # merge the two sorted halves
    merged_arr = merge(left_half, right_half)

    return merged_arr


def merge(left, right):
    merged = []
    left_idx = 0
    right_idx = 0

    # merge the two sorted halves into one array
    while left_idx < len(left) and right_idx < len(right):
        if left[left_idx] < right[right_idx]:
            merged.append(left[left_idx])
            left_idx += 1
        else:
            merged.append(right[right_idx])
            right_idx += 1

    # add the remaining elements to the merged array
    merged.extend(left[left_idx:])
    merged.extend(right[right_idx:])

    return merged


# test
arr = [5, 2, 1, 3, 6, 4]
sorted_arr = merge_sort(arr)
print(sorted_arr)
